### PR TITLE
fix(language-service): resolve the variable from the template context first

### LINKED ISSUE #670: The index of a ngFor is not recongnized as a number again
Following this issue [#514 ](https://github.com/angular/vscode-ng-language-service/issues/514)
It looks like the bug came back on version 0.900.15 (wich i'm currently using).

I have this really simple piece of code : 

```html
<div fxLayout="row" fxLayoutGap="20px grid" fxLayoutAlign="space-evenly center" fxFlex style="text-align: center;">
    <div *ngFor="let table of title; let i = index" style='cursor: pointer;'>
        <div *ngIf='table.length == 0' style='text-align: center;'> Aucune table n'est disponnible dans cette base de donnée </div>
        <div *ngIf='table.length != 0'> <img src='assets/images/table.png' width="50px"><br /> {{table}} </div>
        <div *ngIf="i == 7" class='break'></div>
    </div>
</div>
```

And i have the following error from the langage service :
" Expected the operants to be of similar type or any ng(0)" on this part "*ngIf="i == 7" event thought it's for sure a number.

When i disable the angular Language Service i don't have the error anymore.

Thanks in advance  for your help :)

_edit ayazhafiz_: code formatting

### PR DESCRIPTION
**Pull request by ivanwonder**
fixed: https://github.com/angular/vscode-ng-language-service/issues/670

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/language-service/test/project/app/parsing-cases.ts [hunk 1]
```diff
@@ -170,6 +170,8 @@ export class TemplateReference {
   primitiveIndexType: {[name: string]: string} = {};
   anyValue: any;
   optional?: string;
+  // Use to test the `index` variable conflict between the `ngFor` and component context.
+  index?: string;
```
Comment (kyliau) [inline packages/language-service/test/project/app/parsing-cases.ts]: I think it'd be more obvious if index is just `null`?

File: packages/language-service/src/expression_diagnostics.ts [hunk 2]
```diff
@@ -91,9 +91,13 @@ function getVarDeclarations(
       continue;
     }
     for (const variable of current.variables) {
-      let symbol = info.members.get(variable.value);
+      // Try to resolve the variable from directive context first, then the component context.
+      let symbol = getVariableTypeFromDirectiveContext(variable.value, info.query, current);
```
Comment (kyliau) [inline packages/language-service/src/expression_diagnostics.ts]: consider consolidating these three statements to
```
const symbol = getVariableTypeFromDirectiveContext(...) || info.member.get(...) || info.query.getBuiltinType(...);
```

**Review by kyliau — CHANGES_REQUESTED** [review]

File: packages/language-service/test/diagnostics_spec.ts [hunk 3]
```diff
@@ -986,4 +986,13 @@ describe('diagnostics', () => {
       expect(content.substring(start !, start ! + length !)).toBe(expression);
     }
   });
+
+  it('should not report error if template context conflict with component context', () => {
+    mockHost.override(TEST_TEMPLATE, `
+      <div *ngFor="let h of heroes; let i = index">
```
Comment (kyliau) [inline packages/language-service/test/diagnostics_spec.ts]: Could you please add a comment here to call out `index` refers to the directive context and not the component context? Thank you.

**Review by kyliau — APPROVED** [review]

File: packages/language-service/src/expression_diagnostics.ts [hunk 4]
```diff
@@ -91,10 +91,10 @@ function getVarDeclarations(
       continue;
     }
     for (const variable of current.variables) {
-      let symbol = info.members.get(variable.value);
-      if (!symbol) {
-        symbol = getVariableTypeFromDirectiveContext(variable.value, info.query, current);
-      }
+      // Try to resolve the variable from directive context first, then the component context.
+      let symbol = getVariableTypeFromDirectiveContext(variable.value, info.query, current) ||
+          info.members.get(variable.value) || info.query.getBuiltinType(BuiltinType.Any);
```
Comment (ivanwonder) [inline packages/language-service/src/expression_diagnostics.ts]: @kyliau I think it's wrong to get the type of the variable from component context.
```
*ngIf="let a = b.c.d()"
```
 the RHS of a let expression cannot be an expression

File: packages/language-service/test/diagnostics_spec.ts (hunk 3 above)
Comment (ivanwonder) [reply packages/language-service/test/diagnostics_spec.ts]: Maybe it's no need to add the test case for it. the `index` only can refer to the directive context.

File: packages/language-service/src/expression_diagnostics.ts (hunk 4 above)
Comment (ayazhafiz) [reply packages/language-service/src/expression_diagnostics.ts]: +1

File: packages/language-service/test/diagnostics_spec.ts (hunk 3 above)
Comment (ayazhafiz) [reply packages/language-service/test/diagnostics_spec.ts]: There should already be a test for this when checking the type of an ngFor export.

**Review by ayazhafiz — APPROVED** [review]

File: packages/language-service/src/expression_diagnostics.ts (hunk 4 above)
Comment (ivanwonder) [reply packages/language-service/src/expression_diagnostics.ts]: removed the component context.

File: packages/language-service/test/diagnostics_spec.ts (hunk 3 above)
Comment (ivanwonder) [reply packages/language-service/test/diagnostics_spec.ts]: Just add `index` to the component context.

**Pull request closed without merging**